aiida/tools/importexport/dbexport/__init__.py

```diff
@@ -325,108 +407,122 @@ def export_tree(
 
-    export_data = dict()
+    if entries_to_add:
+        progress_bar = get_progress_bar(total=len(entries_to_add), disable=silent)
+
+    export_data = defaultdict(dict)
     entity_separator = '_'
     for entity_name, partial_query in entries_to_add.items():
 
-        foreign_fields = {
-            k: v
-            for k, v in all_fields_info[entity_name].items()
-            # all_fields_info[model_name].items()
-            if 'requires' in v
-        }
+        progress_bar.set_description_str('Exporting {}s'.format(entity_name), refresh=False)
+        progress_bar.update()
+
+        foreign_fields = {k: v for k, v in all_fields_info[entity_name].items() if 'requires' in v}
 
         for value in foreign_fields.values():
             ref_model_name = value['requires']
             fill_in_query(partial_query, entity_name, ref_model_name, [entity_name], entity_separator)
 
         for temp_d in partial_query.iterdict():
-            for k in temp_d.keys():
+            for key in temp_d:
                 # Get current entity
-                current_entity = k.split(entity_separator)[-1]
+                current_entity = key.split(entity_separator)[-1]
 
                 # This is a empty result of an outer join.
                 # It should not be taken into account.
-                if temp_d[k]['id'] is None:
+                if temp_d[key]['id'] is None:
                     continue
 
-                temp_d2 = {
-                    temp_d[k]['id']:
+                export_data[current_entity].update({
+                    temp_d[key]['id']:
                     serialize_dict(
-                        temp_d[k], remove_fields=['id'], rename_fields=model_fields_to_file_fields[current_entity]
+                        temp_d[key], remove_fields=['id'], rename_fields=model_fields_to_file_fields[current_entity]
                     )
-                }
-                try:
-                    export_data[current_entity].update(temp_d2)
-                except KeyError:
-                    export_data[current_entity] = temp_d2
+                })
+
+    # Close progress up until this point in order to print properly
+    close_progress_bar(leave=False)
 
     #######################################
     # Manually manage attributes and extras
     #######################################
-    # I use .get because there may be no nodes to export
-    all_nodes_pk = list()
-    if NODE_ENTITY_NAME in export_data:
-        all_nodes_pk.extend(export_data.get(NODE_ENTITY_NAME).keys())
+    # Pointer. Renaming, since Nodes have now technically been retrieved and "stored"
+    all_node_pks = node_ids_to_be_exported
 
-    if sum(len(model_data) for model_data in export_data.values()) == 0:
+    model_data = sum(len(model_data) for model_data in export_data.values())
+    if not model_data:
+        msg = 'Nothing to store, exiting...'
         if not silent:
-            print('No nodes to store, exiting...')
+            print(msg)
```

Comment:
I guess this was left over

Reply:
Kind of not. It was intended that this message should also show up in the terminal.

Reply:
Writing to a logger does not necessarily mean it won't end up on the console. In that case one just needs to configure a handler to directs output to stdout.

Reply:
I am aware of that, but then I would need to setup a whole other handler for just these things. But maybe it will be time well spent, when considering also the summaries... Although, I was thinking to have a more condensed version of these go to the logger...

Reply:
So I have made a new handler to simply print all stuff to `sys.stdout`.
However, now is my question - how is this different from `print` (other than it also prints the whole thing to the log), and wouldn't this still have to be under a `if not silent` "clause"? Note, I am logging at the `INFO` level for all the things that should go to the console.
Should I instead use the custom `REPORT` level, or? How would toggling these messages now actually work?

Reply:
Maybe I missed to clarify that I consider `silent` messages and `debug` messages quite different here. The first includes things that would "normally" go to the console to give a nice overview of what's happening, this including the actual progress bar. The latter is for debugging, to be able to see how far in the given `export` function we got until some error occurred, i.e., it's messages for us, not the users.

Reply:
Hi @CasperWA , sorry for the late reply.
My basic idea was: the `silent` flag toggles the log level of the logger - if it is not set, use the default level; otherwise e.g. `ERROR`.

The logger would have a handler that writes to the stdout (and, if desired, also one writing to the AiiDA log file).

I'll now have another look at the code to check your implementation.

Reply:
> Should I instead use the custom `REPORT` level, or? How would toggling these messages now actually work?

Maybe I evaluate the supplied `silent` parameter and set the logging level to `WARNING` if `silent=True`, and otherwise set it to `INFO` - caching the original logging level to be able to reinstate it at the end?

I don't like this though, since it momentarily overrides any level the user/developer may be otherwise expecting.

Reply:
> My basic idea was: the `silent` flag toggles the log level of the logger - if it is not set, use the default level; otherwise e.g. `ERROR`.

See my other comment for a consideration on this.

> The logger would have a handler that writes to the stdout (and, if desired, also one writing to the AiiDA log file).

Yup, set this up already (except the logging to the AiiDA log file, I expected this to be there already since I am using a child Logger of the AiiDA logger).

> I'll now have another look at the code to check your implementation.

I havne't pushed any changes yet, I have it only locally for now.

Reply:
> However, now is my question - how is this different from print (other than it also prints the whole thing to the log)

The difference to `print` is that logging is filtered by the log level.

> , and wouldn't this still have to be under a if not silent "clause"? Note, I am logging at the INFO level for all the things that should go to the console.
Should I instead use the custom REPORT level, or? How would toggling these messages now actually work?

As mentioned above, I suggest you use the `silent` flag to filter out all but the most important log messages (e.g. by setting the log level to `ERROR`).
If the flag isn't set, the default log level should then be `INFO` (I think?), i.e. anything logged with `logger.info` will show up, and anything logged with `logger.debug` won't.

As this example shows, what one would actually want is more than a boolean flag to control the log level. 
My plan is to introduce a verbosity flag in https://github.com/aiidateam/aiida-core/pull/3896 which could then also be used here.

Reply:
Ah right! We just set the handler's logging level based on the `silent` parameters. Done. It doesn't change anything else.

Reply:
> As mentioned above, I suggest you use the `silent` flag to filter out all but the most important log messages (e.g. by setting the log level to `ERROR`).
> If the flag isn't set, the default log level should then be `INFO` (I think?), i.e. anything logged with `logger.info` will show up, and anything logged with `logger.debug` won't.
> 
> As this example shows, what one would actually want is more than a boolean flag to control the log level.
> My plan is to introduce a verbosity flag in #3896 which could then also be used here.

Sure. But all of this can only be respected properly (I think) if we then set the level for the console stream handler according to the `silent` parameter. And even then, it's a bit off, since the target level (if `silent`) is arbitrary _and_ hard-coded.

And while the `verbosity` flag may solve that if implemented here as well, it doesn't solve the cases when not going through the CLI, but instead the Python API directly ...

Reply:
When this thread started a few minutes ago, I wrote a simple example script to demonstrate how the logging can (and I think should) be used. We should not be using `print` for debug or informational messages and we should definitely not be mixing it with logging. I think we should only use logging and it is already set up in AiiDA to be used. I don't know why you think that you should add more handlers, because that is not True. They are already defined and automatically configured and loaded when you load a profile. So if you call a `verdi` command, the logging is enabled. Here is the example script that I think demonstrates how this can be done:
```
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import click

from aiida.cmdline.params import options, types


@click.command()
@options.PROFILE(type=types.ProfileParamType(load_profile=True))
@click.option('--silent', is_flag=True)
def launch(profile, silent):
    """Test CLI script to demonstrate logging."""
    import logging
    from aiida.common.log import AIIDA_LOGGER

    logger = AIIDA_LOGGER.getChild('cli')

    if silent:
        logging.disable(level=logging.CRITICAL)

    logger.debug('this is a debug level message')
    logger.info('this is an info level message')
    logger.warning('this is a warning level message')
    logger.critical('this is a critical level message')


if __name__ == '__main__':
    launch()
```

If you call this with default AiiDA config, you will see:
```
./test_logging.py
05/13/2020 03:01:31 PM <15932> aiida.cli: [WARNING] this is a warning level message
05/13/2020 03:01:31 PM <15932> aiida.cli: [CRITICAL] this is a critical level message
```
If you change the loglevel that changes:
```
verdi config logging.aiida_loglevel DEBUG
05/13/2020 03:06:52 PM <16162> aiida.cli: [DEBUG] this is a debug level message
05/13/2020 03:06:52 PM <16162> aiida.cli: [INFO] this is an info level message
05/13/2020 03:06:52 PM <16162> aiida.cli: [WARNING] this is a warning level message
05/13/2020 03:06:52 PM <16162> aiida.cli: [CRITICAL] this is a critical level message
```
If the `--silent` flag is intended to suppress _everything_ the current example does just that:
```
./test_logging.py --silent
```
now won't print anything.
The only decision then is whether `--silent` should still show errors and or critical messages.

Reply:
> And while the verbosity flag may solve that if implemented here as well, it doesn't solve the cases when not going through the CLI, but instead the Python API directly

You can set the verbosity of the AiiDA logger through the python API of course (this is what the verbosity flag will do).
We just need to document how to do this.

Reply:
> If you call this with default AiiDA config, you will see:
> 
> ```
> ./test_logging.py
> 05/13/2020 03:01:31 PM <15932> aiida.cli: [WARNING] this is a warning level message
> 05/13/2020 03:01:31 PM <15932> aiida.cli: [CRITICAL] this is a critical level message
> ```

Why doesn't this print the `INFO` statement as well?

> The only decision then is whether `--silent` should still show errors and or critical messages.

Fine.

I would need a different formatter though to get the desired output in the console (just `'%(message)s'`).

Reply:
> Why doesn't this print the `INFO` statement as well?

Because the default Python level is `WARNING` and the default in AiiDA is `REPORT` which is a level that we invented and does not even exist in normal PYTHON. Since `REPORT` is in between `INFO` and `WARNING` you don't see info logs by default

Reply:
As I expected then and also referenced in earlier comments, fine.
Then how can I set another formatter without setting up a new handler as well?

Reply:
I basically want to use `echo.echo()` without having to import `echo`, I guess, since this should only be imported and used in the `cmdline` module.

Reply:
Well the following would work:
```
    for handler in AIIDA_LOGGER.handlers:
        handler.setFormatter(logging.Formatter('%(message)s'))
```
but you should reset the changes at the end of the method. When called through `verdi` it is fine, because at the end of the command, the changed formatters don't matter anymore. But if the export methods are called from within a normal interpreter, the logging of all other code that is executed thereafter is also affected. So you should make sure to reset the changes to the logger at the end.

The real underlying problem is that you are affecting normal module code such that its output is such as the CLI interface would like it. But to be honest, this is a bit of a code smell. The CLI is not the only caller and making the code do stuff particularly for that end point is not what we should be doing. In the interest of getting this merged, let's finish this though and just undo the formatting changes at the end of your methods.

Reply:
> The real underlying problem is that you are affecting normal module code such that its output is such as the CLI interface would like it. But to be honest, this is a bit of a code smell. The CLI is not the only caller and making the code do stuff particularly for that end point is not what we should be doing. (...)

But then I would need to set this up specifically in the `log` module, as you have done with certain exceptions for the daemons, etc.